### PR TITLE
Glrm refactoring

### PR DESCRIPTION
**Pull request by st-pasha**
* Split out GlrmLoss into a separate class (in preparation for Mojos);
* Fix multiple warnings and normalized whitespace in GLRM / GLRMModel;
* Removed `H2O.getMCM()` method because it tried to instantiate a utility class `ModelCacheManager`.

---
This change is [Reviewable](https://reviewable.io/reviews/h2oai/h2o-3/285)

### CODE REVIEW

**Comment by st-pasha** [comment]
Note: this is not the final location for the GlrmLoss class -- it will eventually be moved to genmodel, once we have a better understanding of where it should be...

File: h2o-algos/src/main/java/hex/glrm/GLRM.java [hunk 1]
```diff
@@ -83,15 +82,17 @@
...
-    if (_valid != null && _valid.numRows() != _train.numRows()) error("_valid", "_valid must have same number of rows as _train");
+    if (_train.numCols() < 2)
+      error("_train", "_train must have more than one column");
+    if (_valid != null && _valid.numRows() != _train.numRows())
+      error("_valid", "_valid must have same number of rows as _train");
 
-    _ncolY = LinearAlgebraUtils.numColsExp(_train,true);
-    if (_ncolY > MAX_COLS_EXPANDED)
+    _ncolY = LinearAlgebraUtils.numColsExp(_train, true);
+    if (_ncolY > 5000)
```
Comment (mmalohlava) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Ehh, why the constant was replaced by a number?

File: h2o-core/src/main/java/water/Iced.java [hunk 2]
```diff
@@ -31,7 +31,7 @@
  * @see water.Weaver
  * @see water.AutoBuffer
  */
-abstract public class Iced<D extends Iced> implements Freezable<D>, Externalizable {
+abstract public class Iced<D extends Iced> implements Freezable<D> {
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/Iced.java]: Are you sure here? It needs more explanation - this can potentially  break lot of stuff

**Comment by mmalohlava** [comment]
Adding @tomasnykodym since the `Iced` change is strongly affecting his code in DException propagation.

**Comment by wendycwong** [comment]
Pasha:

Took a look at the code.  Just one minor comment.  Please keep the MAX_COLS_EXPANDED constant.  This may change to a different number in the future.

I need to run your change through my REST_API regression test.  Once that passed, I will have no problem with it.

Will let you know.

File: h2o-core/src/main/java/water/Iced.java [hunk 3]
```diff
@@ -116,14 +116,14 @@ final public String toJsonString() {
 //  @Override public D readJSON_impl( AutoBuffer ab ) { return (D)this; }
 
   // Java serializers use H2Os Icing
-  @Override public void readExternal( ObjectInput ois )  throws IOException, ClassNotFoundException {
+  public void readExternal( ObjectInput ois )  throws IOException, ClassNotFoundException {
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/Iced.java]: This is used outside H2O.

File: h2o-core/src/main/java/water/Iced.java [hunk 4]
```diff
@@ -116,14 +116,14 @@ final public String toJsonString() {
...
     int x = ois.readInt();
     byte[] buf = MemoryManager.malloc1(x);
     ois.readFully(buf);
     read(new AutoBuffer(buf));
   }
 
-  @Override public void writeExternal( ObjectOutput oos ) throws IOException {
+  public void writeExternal( ObjectOutput oos ) throws IOException {
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/Iced.java]: This is used outside H2O as well.

File: h2o-core/src/main/java/water/Iced.java (hunk 2 above)
Comment (st-pasha) [inline h2o-core/src/main/java/water/Iced.java]: ok, i'll move this change to a separate pr

**Comment by st-pasha** [comment]
Merging because need to continue working on this class.
The reported build failures concern a missing data.csv file.
Concerns raised by Michal (about changes to Iced) moved into a separate PR

**Comment by mmalohlava** [comment]
Why do you need to merge it? You can continue work on this and push the final commit. Or even create another PR which is based on this branch....

**Comment by wendycwong** [comment]
Pasha:

There is no GLRM failure from your branch.

I have no more questions.  


Wendy

**Pull request merged**